mayastor/src/subsys/mbus/mod.rs

```diff
@@ -0,0 +1,96 @@
+//! Message bus connecting mayastor to the control plane components.
+//!
+//! It is designed to make sending events to control plane easy in the future.
+//!
+//! A Registration subsystem is used to keep moac in the loop
+//! about the lifecycle of mayastor instances.
+
+pub mod mbus_nats;
+pub mod registration;
+
+use async_trait::async_trait;
+use smol::io;
+use spdk_sys::{
+    spdk_subsystem,
+    spdk_subsystem_fini_next,
+    spdk_subsystem_init_next,
+};
+use structopt::StructOpt;
+
+use crate::{
+    core::{MayastorCliArgs, MayastorEnvironment},
+    subsys::mbus::{mbus_nats::NATS_MSG_BUS, registration::Registration},
+};
+
+// wrapper around our MBUS subsystem used for registration
+pub struct MessageBusSubsystem(pub(crate) *mut spdk_subsystem);
+
+impl Default for MessageBusSubsystem {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MessageBusSubsystem {
+    /// initialize a new subsystem that handles the control plane
+    /// message bus registration process
+    extern "C" fn init() {
+        debug!("mayastor mbus subsystem init");
+        let args = MayastorEnvironment::new(MayastorCliArgs::from_args());
+        if let (Some(_), Some(grpc)) = (args.mbus_endpoint, args.grpc_endpoint)
+        {
```

Comment:
:thinking:

Reply:
The registration requires the grpc endpoint so we can give it to moac, so it can only run if both nats and grpc are enabled.